bsp/worker/src/mill/bsp/worker/Utils.scala

```diff
@@ -46,6 +58,73 @@ private object Utils {
     else StatusCode.OK
   }
 
+  def makeBuildTarget(
+      id: BuildTargetIdentifier,
+      depsIds: Seq[BuildTargetIdentifier],
+      bt: BspBuildTarget,
+      data: Option[(String, Object)]
+  ): BuildTarget = {
+    val buildTarget = new BuildTarget(
+      id,
+      bt.tags.asJava,
+      bt.languageIds.asJava,
+      depsIds.asJava,
+      new BuildTargetCapabilities().tap { it =>
+        it.setCanCompile(bt.canCompile)
+        it.setCanTest(bt.canTest)
+        it.setCanRun(bt.canRun)
+        it.setCanDebug(bt.canDebug)
+      }
+    )
+
+    bt.displayName.foreach(buildTarget.setDisplayName)
+    bt.baseDirectory.foreach(p => buildTarget.setBaseDirectory(sanitizeUri(p)))
+
+    for ((dataKind, data) <- data) {
+      buildTarget.setDataKind(dataKind)
+      buildTarget.setData(data)
+    }
+    buildTarget
+  }
+
+  def outputPaths(
+      outDir: os.Path,
+      buildTargetBaseDir: os.Path,
+      topLevelProjectRoot: os.Path
+  ): Seq[OutputPathItem] = {
+    val output = new OutputPathItem(
+      // Spec says, a directory must end with a forward slash
+      sanitizeUri(outDir) + "/",
+      OutputPathItemKind.DIRECTORY
+    )
+
+    def ignore(path: os.Path): Boolean = {
+      path.last.startsWith(".") ||
+      path.last == "out" ||
+      path.last == "target"
+    }
+
+    def projectsRootPaths = os.walk(topLevelProjectRoot, ignore).collect {
+      case p if p.endsWith(os.RelPath("build.sc")) => p / os.up
+    }
+    def outputPathItem(path: os.Path) =
+      new OutputPathItem(
+        // Spec says, a directory must end with a forward slash
+        sanitizeUri(path) + "/",
+        OutputPathItemKind.DIRECTORY
+      )
+    def additionalExclusions = projectsRootPaths.flatMap { path =>
+      Seq(
+        outputPathItem(path / ".idea"),
+        outputPathItem(path / "out"),
+        outputPathItem(path / ".bsp"),
+        outputPathItem(path / ".bloop")
+      )
+    }
```

Comment:
I'm not convinces excluding hardcoded directories is the right approach for a generic protocol. Esp. since some of these directories aren't even created or managed by us.